azure-pipelines/main.yml

```diff
@@ -0,0 +1,108 @@
+# Minimal Build pipeline
+# This allow us to generate wheels and upload that to a feed.
+
+variables:
+  artifactFeed:
+
+  wheels.dir: /tmp/tensorflow_pkg
+  bazel.startup.opts: --batch_cpu_scheduling --io_nice_level 7
+  bazel.build.opts: --jobs 1 --local_ram_resources=HOST_RAM*.6
+  numpy.version: '<1.19.0'
+
+name: $(BuildDefinitionName)_$(SourceBranchName)_$(BuildID)
+stages:
+  - stage: Build
+    jobs:
+      - job: bazelBuild
+        timeoutInMinutes: 1440
+        cancelTimeoutInMinutes: 35
+        strategy:
+          matrix:
+            linux_python38:
+              imageName: 'ubuntu-20.04'
+              python.version: '3.8'
+              use.bazel.version: 3.1.0
+              pool.name: 'ADORayTests'
+#            macos1015_python38:
+#              imageName: 'macOS-10.15'
+#              python.version: '3.8'
+#              use.bazel.version: 3.1.0
+        displayName: Bazel build
+        pool:
+#          vmImage: $(imageName)
+          name: $(pool.name)
+        steps:
+          - template: templates/python-dev.yml
+          - template: templates/check-bazel.yml
+          - bash: |
+              pushd $BUILD_SOURCESDIRECTORY
+              yes "" | ./configure
+              popd
+            displayName: "Configure a default TF2 build"
+          - template : templates/wheel-version.yml
+          - bash: |
+              # Doesn't work if installed in just the env, for includes
+              # specific to TF 2.3.0 -- restrict numpy version.
+              python -m pip install -U "numpy${NUMPY_VERSION}"
+
+              source $PENV_HOME/bin/activate
+              echo switched to venv python: $(which python)
+              gcc --version
+              bazelisk --version
+
+              python -m pip install -U "numpy${NUMPY_VERSION}"
+              python -m pip install -U keras_preprocessing --no-deps
+              python -m pip install -U keras_applications --no-deps
+
+              # Retry for spurious failures. Bazel should cache successful steps.
+              for i in 1 2; do
+
+                echo Compilation attempt $i...
+
+                bazelisk $BAZEL_STARTUP_OPTS build $BAZEL_BUILD_OPTS //tensorflow/tools/pip_package:build_pip_package
+                compile_result="$?"
+                if [[ compile_result == 0 ]]; then
+                  break
+                fi
+              done
+
+              if [[ compile_result != 0 ]]; then
+                echo Compilation failed after $i attempts.
+              fi
+
+              exit $compile_result
+            displayName: 'Bazel build'
+            timeoutInMinutes: 1440
```

Comment:
you have the timeout for this task and the following one set equal to the timeout of the whole job.
I'm not sure if that is a problem or not, but the practice I have followed goes like this:
- Set the job timeout to the worst case I can imagine (you already did this)
- Keep all tasks in the job without timeout if possible
- Use the timeout of task if and only if I want to make the pipeline fail fast in an initial step that should run fast but tend to be flaky (like installing dependencies)

Again, I'm not sure what you have is current or not, but it could, so I'm sharing the approach that has worked for me so far.